### DISCUSSION TITLE
SwapPool Interface Design

### DISCUSSION THREAD

**Discussion by WillRuddick** [grassrootseconomics/sarafu.network#89] [Ideas]
Develop a SwapPool interface on Sarafu.Network/swap 
that allows someone to 
Define a Pool, Deposit into a pool and Exchange

**Create**
- which tokens can go in it and for each there may be a cap - Address and cap (max quantity)
- Define relative pricing assume 1:1
- Define fees: assume 0

**Deposit:** Add their assets to the pool (given the rules above)
**Exchange** assets (given the rules above): Add a voucher and specify which other voucher you want in exchange.

Consider a get quote function for checking prices before executing any transfers. (control for potential slippage)
Note that the Price and be pushed to an oracle (like Uniswap can handle bonding curves if needed) (by default static rate)

**MVP**
Basic functions and visibility to reach the swap-pool on sarafu.network/swap

**Notes:**

Keep a registry of these contracts up to date (sync) (on chain- and off chain)
and a basic way to lookup existing pools (similar to exploring vouchers)
Consider a graph that holds the nodes (vouchers) and edges (pools) in order to perform offline operations - like shortest paths.
Consider a market place view where all other vouchers (products) are shown that can be reached by this graph of vouchers and pools. The avalaible products to look for would be all those paths from the users holdings to all other vouchers.
Consider a visualization of this graph (similar to that of transfers on (viz.sarafu.network)

**use cases**

1. upgrading contracts
someone makes a new voucher, puts it in the pool and specifices that the old one can be swapped
old contracts (with replacements) can be flagged

2. Specifying that someone wants to exchange their vouchers (products) for an amount of another asset (like cUSD). // vending machine 

https://github.com/nolash/erc20-pool here is the pool contract we've been working on. 
https://github.com/nolash/erc20-limiter this is the limiter as well (for capping indexed vouchers)

**Comment by sirpy** [comment]
Can you explain exactly how does it differ from standard AMM? lets say uniswap v3.
What kind of features are you looking for?
Cap - If you use a v3 pool with very narrow price range, you automatically get a cap. because once it is out of the price range the pool holds a single token and it is no longer possible to buy the other token.

**Comment by WillRuddick** [comment]
> Can you explain exactly how does it differ from standard AMM? lets say uniswap v3. What kind of features are you looking for? Cap - If you use a v3 pool with very narrow price range, you automatically get a cap. because once it is out of the price range the pool holds a single token and it is no longer possible to buy the other token.

The idea is to offer a simple static rate between vouchers as well as an option for an oracle so you can have dynamic rates if needed. Also to be able to have more than 2 assets in the pool (all are swapable)

We also want to do away with the pool token if possible. 

Uniswap vs is nice - but I've yet to get our demurrage contracts working on there @kamikazechaser can say more

**Comment by sirpy** [comment]
rebased tokens should work with v3, but the liquidity provider will incur the lose.
i'd generally wouldnt recommend creating this kind of non standard pools as it will be locked out of the ecosystem tooling.
if you need oracle based price you can take out the liquidity from the v3 pool and add it at another price range when the price changes. If you are not expecting the price to not change much it makes more sense.
The only problem is multiple tokens, but it can be solved if there's simply  a swap path, a->b->c.

**Comment by WillRuddick** [comment]
@sirpy please feel free to get the voucher contracts working with uniswap or help put us in touch with someone that can help.

**Comment by sirpy** [comment]
@WillRuddick 
1. do you have any deployed ones? addresses?
2. can you send me some vouchers?
3. link to the contracts?

**Comment by WillRuddick** [comment]
![6b257e50-ab55-4e5a-aa33-d0aee64f671a](https://github.com/grassrootseconomics/sarafu.network/assets/12640165/5970b34b-f367-4dac-a054-da8437068ee6)
(ux drafting session 20Nov2023 @williamluke4 )

**Comment by williamluke4** [comment]
> @WillRuddick
> 
> 1. do you have any deployed ones? addresses?
> 2. can you send me some vouchers?
> 3. link to the contracts?

@sirpy You can create your own voucher here https://sarafu.network/publish let me know if you need some CELO

**Comment by walzph** [comment]
I recently got involved with GrassrootsEconomics and @WillRuddick just introduced me to this discussion. So here are my thoughts:

After looking into the [SwapPool](https://github.com/nolash/erc20-pool) contract I understand why some are more convinced by that. No matter how you tweak the concentrated liquidity parameters of Uniswap v3 as proposed by @sirpy, the price is still determined based on a constant product bonding curve. Which makes total sense if the objective is to generate market equilibrium between openly accessible assets. 

But my question would be: Is this even a desirable mechanism for a voucher marketplace?
Probably most vouchers are quite limited in their supply and/or their redeemability, so some trades will have a significant price impact on the pool. And because there won't be any arbitrage traders (you probably won't find many exchanges for these vouchers) this imbalance might not cure itself. 
In contrast, using an external quoter and a token limiter could allow for more customizable and stable pricing models.
Of course, this approach compromises on decentralization, requires users to trust the pool owner and might even negatively influence optimal resource allocation in the long run.

As a final thought I want to mention [Balancer](https://balancer.fi/) and [Symmetric](https://celo.symm.fi/#/) (a fork on Celo). This protocol enables multi-token pools with weights assigned to each asset (max. 8 token though). This gives more flexibility and control in pricing mechanisms compared to other standard DeFi protocols. Would be interested in what you think about them.

**Comment by WillRuddick** [comment]
> As a final thought I want to mention [Balancer](https://balancer.fi/) and [Symmetric](https://celo.symm.fi/#/) (a fork on Celo). This protocol enables multi-token pools with weights assigned to each asset (max. 8 token though). This gives more flexibility and control in pricing mechanisms compared to other standard DeFi protocols. Would be interested in what you think about them.

![image](https://github.com/grassrootseconomics/sarafu.network/assets/12640165/46d3b63b-817e-4872-8317-d73a6ab2f59e)

The idea of having static rate pools (exchanges) seems to be missing with these (like most DEXs).

**Comment by Avkubs** [comment]
I support this result

**Comment by WillRuddick** [comment]
**Pool Publishing (like Create Voucher) Defining a pool (publishing the contract) - like the stepper for Voucher Creation.** 
#1. Define the vouchers allowed in the pool - any erc20 address. 
#2. Define the amount of each voucher allowed in the pool (up to MAX integer)
#3. Define the relative value of each voucher (default 1 (they are all equal))
#4. Define any transaction fees and where they go (like a sink account - defult to use my own address)

Create Pool ... Sign

**Deposit Tokens into the Pool**
Select Token and amount - send - Sign

**Exchange Vouchers in the Pool**
Select input Token and amount, Select avalible output token (and amount - adjust to input)
Select target wallet (default is your own for a swap vs another for a purchase)
Sign

**Auto Routing**
Given a sender address and their default token - calculate the 'credit' someone has in their wallet to exchange there.
